### ISSUE TITLE
Consider refactoring event publishing in AutorecoveringConnection

### ISSUE THREAD

**Issue by antiduh** [rabbitmq/rabbitmq-dotnet-client#516]
Split from #512 .

`AutorecoveringConnection` ("ARC") implements 6 events as part of the `IConnection` interface:

```
        event EventHandler<CallbackExceptionEventArgs> CallbackException;

        event EventHandler<EventArgs> RecoverySucceeded;

        event EventHandler<ConnectionRecoveryErrorEventArgs> ConnectionRecoveryError;

        event EventHandler<ConnectionBlockedEventArgs> ConnectionBlocked;

        event EventHandler<ShutdownEventArgs> ConnectionShutdown;

        event EventHandler<EventArgs> ConnectionUnblocked;
```

ARC uses an internal `Connection` instance (`m_delegate`) for the majority of its implementation, wrapping it to provide the autorecovery semantics. As a result of this, ARC replaces the `m_delegate` instance whenever recovery is triggered and succeeds.

Below are the two existing call sites in ARC that assign the `m_delegate` variable:

```
        private bool RecoverConnectionDelegate()
        {
            while( !ManuallyClosed )
            {
                try
                {
                    var fh = endpoints.SelectOne(m_factory.CreateFrameHandler);
                    m_delegate = new Connection( m_factory, false, fh, this.ClientProvidedName );
                    return true;
                }
            ...
        }

        private void Init(IFrameHandler fh)
        {
            m_delegate = new Connection(m_factory, false,
                fh, this.ClientProvidedName);

```

ARC implements several of its events simply by forwarding its event subscribers to the relevant event handler in `Connection`, since it is acting as a wrapper. However, since the recovery process replaces the instance behind `m_delegate`, ARC has to keep a copy of the event subscribers in a separate collection, to fill back into the new `m_delegate` object.

For instance, consider the flow around the `ConnectionBlocked` event.

Here you can see it forwarding event subscribers to the `m_delegate`, as well as saving a copy for later:

```
        public event EventHandler<ConnectionBlockedEventArgs> ConnectionBlocked
        {
            add
            {
                lock (m_eventLock)
                {
                    m_recordedBlockedEventHandlers += value;
                    m_delegate.ConnectionBlocked += value;
                }
            }
            remove
            {
                lock (m_eventLock)
                {
                    m_recordedBlockedEventHandlers -= value;
                    m_delegate.ConnectionBlocked -= value;
                }
            }
        }

```

And then, after the `m_delegate` instance is replaced in `RecoverConnectionDelegate()`, you can see the subscribers being re-added in `PerformAutomaticRecovery()`:

```
        private void PerformAutomaticRecovery()
        {
            ESLog.Info("Performing automatic recovery");
            lock (recoveryLockTarget)
            {
                if (RecoverConnectionDelegate())
                {
                    RecoverConnectionShutdownHandlers();
                    RecoverConnectionBlockedHandlers();
                    RecoverConnectionUnblockedHandlers();
```

This design is needlessly complicated (requiring extra collections, locks, etc), results in event handlers getting callbacks from objects they never subscribed to (and thus have the wrong `sender` value in their handlers), complicates event registration and handler lifetime, and is fighting against C#'s language features instead of using them.

Instead: declare vanilla event handlers; add explicit subscribers to `m_delegate`'s events when it is constructed or replaced; and forward event invocations from `m_delegate` to ARC's own events.

```
// Declare vanilla event handlers instead of the 'filtering' design:

        public event EventHandler<EventArgs> RecoverySucceeded;

        public event EventHandler<ConnectionRecoveryErrorEventArgs> ConnectionRecoveryError;

        public event EventHandler<CallbackExceptionEventArgs> CallbackException;

        public event EventHandler<ConnectionBlockedEventArgs> ConnectionBlocked;

        public event EventHandler<ShutdownEventArgs> ConnectionShutdown;

        public event EventHandler<EventArgs> ConnectionUnblocked;

// Subscribe to m_delegate's events as necessary:

        private void SubscribeDelegateEvents( Connection conn )
        {
            conn.CallbackException += Connection_CallbackException;
            conn.ConnectionBlocked += Connection_ConnectionBlocked;
            conn.ConnectionRecoveryError += Connection_ConnectionRecoveryError;
            conn.ConnectionShutdown += Connection_ConnectionShutdown;
            conn.ConnectionUnblocked += Connection_ConnectionUnblocked;
            conn.RecoverySucceeded += Connection_RecoverySucceeded;
        }

        private void UnsubscribeDelegateEvents( Connection conn )
        {
            if( conn == null ) 
            { 
                return;
            }

            conn.CallbackException -= Connection_CallbackException;
            conn.ConnectionBlocked -= Connection_ConnectionBlocked;
            conn.ConnectionRecoveryError -= Connection_ConnectionRecoveryError;
            conn.ConnectionShutdown -= Connection_ConnectionShutdown;
            conn.ConnectionUnblocked -= Connection_ConnectionUnblocked;
            conn.RecoverySucceeded -= Connection_RecoverySucceeded;
        }

        private bool RecoverConnectionDelegate()
        {
            while( !ManuallyClosed )
            {
                try
                {
                    var fh = endpoints.SelectOne(m_factory.CreateFrameHandler);

                    // Tempt any exceptions that will occur, leaving m_delegate (and its events) alone if an exception
                    // comes out of the constructor.               
                    var tempConnection = new Connection( m_factory, false, fh, this.ClientProvidedName );

                    // Connection attempt succeeded. Rewire the events and save the new connection.
>>>                 UnsubscribeDelegateEvents( m_delegate );
>>>                 SubscribeDelegateEvents( tempConnection );
                    m_delegate = tempConnection;
                    return true;
                }
                ...
        }

        private void Init(IFrameHandler fh)
        {
            m_delegate = new Connection(m_factory, false,
                fh, this.ClientProvidedName);

>>>         SubscribeDelegateEvents( m_delegate );
            ...
        }


// Forward events from your handlers for m_delegate to your own subscribers, replacing 
// the sender so that subscribers get the right value and can reasonably use it:

        private void Connection_RecoverySucceeded( object sender, EventArgs e )
        {
            RecoverySucceeded?.Invoke( this, e );
        }

        private void Connection_ConnectionUnblocked( object sender, EventArgs e )
        {
            ConnectionUnblocked?.Invoke( this, e );
        }

        private void Connection_ConnectionShutdown( object sender, ShutdownEventArgs e )
        {
            ConnectionShutdown?.Invoke( this, e );
        }

        private void Connection_ConnectionRecoveryError( object sender, ConnectionRecoveryErrorEventArgs e )
        {
            ConnectionRecoveryError?.Invoke( this, e );
        }

        private void Connection_ConnectionBlocked( object sender, ConnectionBlockedEventArgs e )
        {
            ConnectionBlocked?.Invoke( this, e );
        }

        private void Connection_CallbackException( object sender, CallbackExceptionEventArgs e )
        {
            CallbackException( this, e );
        }

```

There may be some details I've left out for brevity or because I haven't spotted them yet; I apologize, I'm new to the client's codebase.

My intent with these changes are to simplify the event handling process for ARC and Connection, while also fixing bugs; I do not intend to introduce any breaking changes.

**Comment by antiduh** [comment]
@michaelklishin, in #512, you commented this:

> I don't see how the provided example would make recovery possible since after recovery the delegate is replaced with a brand new object with no prior state or configuration applied to it. So it may be "less janky" but completely fails to do what `AutorecoveringConnection` was created to do?

Making breaking changes was not my intention; now that I've reviewed the code in ARC more, I have a better understanding of the lifetime of the `Connection` instance behind `m_delegate`, and I see the problem I introduced. I'd ask that you try to interpret my intent, and point out any bugs I inadvertently introduce. 

Specifically, however: "[it] is replaced with a brand new object with no prior state or configuration applied to it". That's what the existing code already does. The rest of the recovery process is involved in restoring configuration and state, and I was meaning to not disturb that.

**Comment by michaelklishin** [comment]
@antiduh I'm stretched a bit thin right now but we will review and consider your suggestions. It is usually easier to do when they come in the form of pull requests, for two reasons:

 * We can catch at least some breaking changes via automated tests
 * A pull request has to be complete to demonstrate the entire idea, even if somewhat incomplete

It is somewhat difficult for me to interpret an isolated snippet in an issue. I understand that producing a PR is more effort for you 🙏

**Comment by antiduh** [comment]
Absolutely understood, I was planning on putting together a pull request pending your approval. I just don't want to suggest something, do the work for it, and then in the pr have you guys say "this doesn't make sense".

**Comment by lukebakken** [comment]
This issue has not received attention in almost a year and will be closed soon as we prepare for version `6.0`.

**Comment by lukebakken** [comment]
Closing due to inactivity. I added the `next-gen-todo` label for us to return to this in the future.
